pkg/virtctl/set/set.go

```diff
@@ -0,0 +1,156 @@
+			return s.Run(cmd, args)
+		},
+	}
+	cmd.Flags().IntVar(&s.cpuSocketsCount, "cpu", 0, "Number of CPUs (sockets) to set")
+	cmd.Flags().StringVar(&s.gustMemorySize, "memory", "", "Memory size to set (e.g., 2Gi)")
+	cmd.Flags().DurationVar(&s.timeout, "timeout", 10*time.Second, "Timeout duration (e.g., 10s, 2m)")
+	cmd.SetUsageTemplate(templates.UsageTemplate())
+	return cmd
+}
+
+func setUsage() string {
+	return `  # Set the number of CPUs (sockets) to 2 and memory to 1Gi for VirtualMachine 'myvm':
+  {{ProgramName}} set myvm --cpu=2 --memory=1Gi`
+}
+
+func (s *Set) Run(cmd *cobra.Command, args []string) error {
+	vmName := args[0]
+	namespace, _, err := s.clientConfig.Namespace()
+	if err != nil {
+		return err
+	}
+
+	if s.cpuSocketsCount < 0 {
+		return fmt.Errorf("invalid CPU count: %d; must be a positive integer", s.cpuSocketsCount)
+	}
+
+	if s.gustMemorySize != "" {
+		quantity, err := resource.ParseQuantity(s.gustMemorySize)
+		if err != nil {
+			return fmt.Errorf("invalid memory size: %s", s.gustMemorySize)
+		}
+		if quantity.IsZero() || quantity.Sign() == -1 {
+			return fmt.Errorf("memory size must be greater than zero")
+		}
+	}
+
+	if s.cpuSocketsCount <= 0 && s.gustMemorySize == "" {
+		return fmt.Errorf("at least one of --cpu or --memory must be set")
+	}
+
+	virtCli, err := kubecli.GetKubevirtClientFromClientConfig(s.clientConfig)
+	if err != nil {
+		return err
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), s.timeout)
+	defer cancel()
+
+	vm, err := virtCli.VirtualMachine(namespace).Get(ctx, vmName, k8smetav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	messages, err := s.setCPUAndMemory(ctx, virtCli, vm)
+	if err != nil {
+		return err
+	}
+
+	for _, message := range messages {
+		cmd.Printf("%s\n", message)
+	}
+	return nil
+}
+
+func (s *Set) setCPUAndMemory(ctx context.Context, virtCli kubecli.KubevirtClient, vm *v1.VirtualMachine) ([]string, error) {
+	domain := vm.Spec.Template.Spec.Domain
+	messages := []string{}
+
+	if s.cpuSocketsCount > 0 {
+		if domain.CPU == nil {
+			domain.CPU = &v1.CPU{}
+		}
+		domain.CPU.Sockets = uint32(s.cpuSocketsCount)
+		messages = append(messages, fmt.Sprintf("Successfully set %d CPUs for %s", s.cpuSocketsCount, vm.Name))
+	}
+
+	if s.gustMemorySize != "" {
+		quantity, err := resource.ParseQuantity(s.gustMemorySize)
+		if err != nil {
+			return nil, fmt.Errorf("invalid memory size parameter: %w", err)
+		}
+		if quantity.IsZero() {
+			return nil, fmt.Errorf("memory size parameter must be greater than zero")
+		}
+		if domain.Memory == nil {
+			domain.Memory = &v1.Memory{}
+		}
+		domain.Memory.Guest = &quantity
+		messages = append(messages, fmt.Sprintf("Successfully set memory to %s for %s", s.gustMemorySize, vm.Name))
+	}
+
+	patchPayload, err := patch.New(patch.WithReplace("/spec/template/spec/domain", domain)).GeneratePayload()
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate patch payload: %w", err)
+	}
```

Comment:
this patch needs a test operation to ensure we don't mutate something unexpected.

Something like this should fix it.

```
        patchSet := patch.New(patch.WithTest("/spec/template/spec/domain", domain)) 
        patchSet.AddOption(patch.WithReplace("/spec/template/spec/domain", domain))
        patchPayload, err := patchSet.GeneratePayload()
```

Reply:
You have a point, but if i am down to performing 2 operations in the patchSet then i prefer it to be 2 small specific operations of replacing the cpu and memory, instead of the entire domain test and domain replace.

to be honest i am not sure what you suggested will work, since test operation will perform a check to make sure the domain object is equal to the current vm.domain but it would definitely fail since I modified some of the fields and sometimes i even create new ones if they don't exist.